open-webui: build failures

### Nixpkgs version

- Stable (25.05)

### Describe the bug

Trying to get the latest version of open-webui and am hitting a bunch of build failures. There's an issue with several langchain tests requiring a bash shell, conflicting python version requirements, and other various issues when it gets built, and since there's no cached version with Hydra since it's unfree, this means you have to build from source each time. Working around these with overlays but it's janky.

### Steps to reproduce

# Steps to Reproduce

The following steps reproduce the build failures encountered when
attempting to build `open-webui` on NixOS (both with and without
overlays).

These steps assume:
- NixOS 24.11 or unstable (any recent revision)
- flakes enabled
- Python 3.13 is the default interpreter in nixpkgs
- `open-webui` version 0.6.36 (current in nixpkgs)

---

## 1. Create a minimal flake using NixOS + home-manager + open-webui

**flake.nix:**

```nix
{
  description = "Minimal reproduction of open-webui build failures";

  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
    home-manager.url = "github:nix-community/home-manager";
    home-manager.inputs.nixpkgs.follows = "nixpkgs";
  };

  outputs = { self, nixpkgs, home-manager, ... }:
    let
      system = "x86_64-linux";
    in {
      nixosConfigurations.repro-host = nixpkgs.lib.nixosSystem {
        inherit system;
        modules = [
          home-manager.nixosModules.home-manager

          {
            services.open-webui.enable = true;

            # Optional: allow unfree so open-webui enables MongoDB
            nixpkgs.config.allowUnfree = true;

            # Minimal config for trusted-users (builds can run as non-root)
            nix.settings.trusted-users = [ "root" ];

            # Standard options
            networking.hostName = "repro-host";
            system.stateVersion = "24.11";
          }
        ];
      };
    };
}
````

---

## 2. Attempt to build the system

Run:

```bash
sudo nixos-rebuild switch --flake .#repro-host
```

---

## 3. Observe failures in Python dependencies

During the build, the following Python packages consistently fail:

### `langchain`

* Fails due to tests invoking `/bin/bash` inside sandbox

### `langchain-community`

* Fails during test collection (21+ errors)

### `fpdf2`

* Fails due to deterministic image hash comparisons differing from upstream

---

## 4. Observe that `open-webui` fails even with dependent test suites disabled

Even when adding overlays that disable test phases for all three Python
dependencies, `open-webui` still fails during `pythonCatchConflictsPhase`
with:

* duplicate `langchain` derivations in closure
* conflict hooks aborting the build

---

## 5. If running the build as a non-root user, observe trust issues

Running:

```bash
nix build nixpkgs#open-webui
```

may emit:

```
[INFO] You are not trusted by store uri: daemon
```

which disables substituters and forces all builds to occur locally.

---

## 6. If `allowUnfree = true`, observe MongoDB builds locally

Because `mongodb` is marked unfree in nixpkgs, Hydra does not provide a
binary substitute. This causes the build to attempt:

```
/nix/store/...-mongodb-x.y.z.drv
```

which significantly increases build time and contributes to multiple
cascading failures.

---

# Result

Following the above steps reproducibly leads to:

1. Test failures in:

   * `langchain`
   * `langchain-community`
   * `fpdf2`

2. Failure of `open-webui` to pass `pythonCatchConflictsPhase` due to
   closure duplication of the `langchain` package.

3. Local rebuilds of MongoDB due to `unfree` license and no Hydra cache.

These steps should reproduce the failures consistently when building
`open-webui` from nixpkgs on NixOS.


### Expected behaviour

Should just build with no issues.

### Screenshots

_No response_

### Relevant log output

```console
# Relevant Log Output

Below is a collection of the relevant build failures encountered while
attempting to build `open-webui` on NixOS, including Python dependency
failures, test failures, and closure conflicts. All identifying system
information has been removed.

---

## 1. `langchain` test failure (`/bin/bash` missing)



error: Cannot build '...-python3.13-langchain-1.0.2.drv'.
Reason: builder failed with exit code 1.

Last 25 log lines:
FileNotFoundError: [Errno 2] No such file or directory: '/bin/bash'

--------------------------- snapshot report summary ----------------------------
22 snapshots passed.
============================= slowest 5 durations ==============================
...
=========================== short test summary info ============================



---

## 2. `langchain-community` multiple test failures



error: Cannot build '...-python3.13-langchain-community-0.3.27.drv'.
Reason: builder failed with exit code 2.

Last 25 log lines:
ERROR tests/unit_tests/agents/test_openai_assistant.py
ERROR tests/unit_tests/agents/test_react.py
ERROR tests/unit_tests/agents/test_tools.py
ERROR tests/unit_tests/callbacks/test_callback_manager.py
ERROR tests/unit_tests/chains/test_api.py
...
!!!!!!!!!!!!!!!!!!! Interrupted: 21 errors during collection !!!!!!!!!!!!!!!!!!!
================ 1 deselected, 19 warnings, 21 errors ==========================



---

## 3. `fpdf2` deterministic output test failures (hash mismatches)



error: Cannot build '...-python3.13-fpdf2-2.8.4.drv'.
Reason: builder failed with exit code 1.

Last 25 log lines:
FAILED test/image/image_types/test_insert_images.py::test_insert_png_monochromatic
FAILED test/image/image_types/test_insert_images.py::test_insert_g4_tiff
FAILED test/image/test_image_info.py::test_get_img_info
FAILED test/table/test_table_with_image.py::test_table_with_qrcode
FAILED test/template/test_flextemplate.py::test_flextemplate_rotation
FAILED test/template/test_flextemplate.py::test_flextemplate_elements
FAILED test/template/test_template.py::test_template_qrcode

= 7 failed, 1371 passed, 6 deselected, 1 xfailed, 7 warnings =



---

## 4. `open-webui` closure conflict: duplicate `langchain`

Even after disabling failing tests, `open-webui` fails during
`pythonCatchConflictsPhase` due to duplicated dependency closures.



error: Cannot build '...-open-webui-0.6.36.drv'.
Reason: builder failed with exit code 1.

Last 25 log lines:
Running phase: pythonCatchConflictsPhase
Found duplicated packages in closure for dependency 'langchain':

langchain 1.0.2 (/nix/store/na996kg1sd1yp076dawks7kivwl5ddma-python3.13-langchain-1.0.2)
dependency chain:
...depending on: ...-python3.13-langchain-1.0.2

langchain 1.0.2 (/nix/store/f3rd6n8vnv7dj0qpr0dyyp1f5anl6ama-python3.13-langchain-1.0.2)
dependency chain:
...depending on: ...-python3.13-langchain-community-0.3.27
...depending on: ...-python3.13-langchain-1.0.2

Package duplicates found in closure. Usually this happens if two
packages depend on different versions of the same dependency.



---

## 5. Additional background errors

### User not trusted (when running build as non-root)



[INFO] You are not trusted by store uri: daemon



This prevented substituters from being used.

### MongoDB being built locally (unfree license)



mongodb is unfree; substituters skipped
these 1 derivations will be built:
/nix/store/...-mongodb-6.x.drv



This significantly increases local build time.

---

# Summary

Several Python dependencies required by `open-webui` fail in their
`checkPhase`, including:

- `langchain`
- `langchain-community`
- `fpdf2`

Additionally, `open-webui` fails due to duplicated `langchain` derivations
during `pythonCatchConflictsPhase`.

These failures prevent `open-webui` from being successfully built on NixOS
without overriding test phases and/or removing conflict hooks.
```

### Additional context

_No response_

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.16.0, NixOS, 25.11 (Xantusia), 25.11.20250814.fbcf476`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.28.4`
 - channels(root): `"nixos-25.05"`
 - nixpkgs: `/nix/store/75d0qr8zvhv921ili0y2fqxa5drk9xpi-source`

### Notify maintainers

@shivaraj-bh @codgician 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+bug%22+-label%3A%226.topic%3A+darwin%22+-label%3A%226.topic%3A+nixos%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/python-modules/langchain-classic/default.nix
+++ b/pkgs/development/python-modules/langchain-classic/default.nix
@@ -1,0 +1,114 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchFromGitHub,
+
+  # build-system
+  hatchling,
+
+  # dependencies
+  langchain-core,
+  langchain-text-splitters,
+  langsmith,
+  pydantic,
+  pyyaml,
+  requests,
+  sqlalchemy,
+
+  # tests
+  blockbuster,
+  cffi,
+  freezegun,
+  langchain-openai,
+  langchain-tests,
+  lark,
+  numpy,
+  packaging,
+  pandas,
+  pytest-asyncio,
+  pytest-dotenv,
+  pytest-mock,
+  pytest-socket,
+  pytest-xdist,
+  pytestCheckHook,
+  requests-mock,
+  responses,
+  syrupy,
+  toml,
+}:
+
+buildPythonPackage rec {
+  pname = "langchain-classic";
+  version = "1.0.0-unstable-2025-11-11";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "langchain-ai";
+    repo = "langchain";
+    # no tagged releases avaialble
+    rev = "3dfea96ec1d2dac4e506d287860ee943c183c9f1";
+    hash = "sha256-U3UllSSa4tFz+nXAP6aNoYceU/xCPbwKSP2F2et+qgQ=";
+  };
+
+  sourceRoot = "${src.name}/libs/langchain";
+
+  build-system = [ hatchling ];
+
+  pythonRelaxDeps = [
+    # Each component release requests the exact latest core.
+    "langchain-core"
+  ];
+
+  dependencies = [
+    langchain-core
+    langchain-text-splitters
+    langsmith
+    pydantic
+    pyyaml
+    requests
+    sqlalchemy
+  ];
+
+  nativeCheckInputs = [
+    blockbuster
+    cffi
+    freezegun
+    langchain-core
+    langchain-openai
+    langchain-tests
+    langchain-text-splitters
+    lark
+    numpy
+    packaging
+    pandas
+    pytest-asyncio
+    pytest-dotenv
+    pytest-mock
+    pytest-socket
+    pytest-xdist
+    pytestCheckHook
+    requests-mock
+    responses
+    syrupy
+    toml
+  ];
+
+  enabledTestPaths = [
+    # integration_tests require network access, database access and require `OPENAI_API_KEY`, etc.
+    "tests/unit_tests"
+  ];
+
+  disabledTests = [
+    # Network access (web.example.com)
+    "test_socket_disabled"
+  ];
+
+  pythonImportsCheck = [ "langchain_classic" ];
+
+  meta = {
+    description = "Classic (0.x) compatibility layer for LangChain";
+    homepage = "https://github.com/langchain-ai/langchain/tree/master/libs/langchain";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ sarahec ];
+  };
+}

--- a/pkgs/development/python-modules/langchain-community/default.nix
+++ b/pkgs/development/python-modules/langchain-community/default.nix
@@ -10,7 +10,7 @@
   aiohttp,
   dataclasses-json,
   httpx-sse,
-  langchain,
+  langchain-classic,
   langchain-core,
   langsmith,
   numpy,
@@ -42,39 +42,30 @@
 
 buildPythonPackage rec {
   pname = "langchain-community";
-  version = "0.3.27";
+  version = "0.4.1";
   pyproject = true;
 
   src = fetchFromGitHub {
     owner = "langchain-ai";
     repo = "langchain-community";
     tag = "libs/community/v${version}";
-    hash = "sha256-rGU8AYe7993+zMAtHLkNiK+wA+UtZnGkUQsOPMtUQ8w=";
+    hash = "sha256-N92YDmej2shQQlktr0veFOKyGFWemFj0hdJIYu1rYSc=";
   };
 
   sourceRoot = "${src.name}/libs/community";
 
   build-system = [ pdm-backend ];
 
+  # Only needed for mixed python 3.12/3.13 builds
   pythonRelaxDeps = [
-    # Each component release requests the exact latest langchain and -core.
-    # That prevents us from updating individual components.
-    "langchain"
-    "langchain-core"
     "numpy"
-    "pydantic-settings"
-    "tenacity"
-  ];
-
-  pythonRemoveDeps = [
-    "bs4"
   ];
 
   dependencies = [
     aiohttp
     dataclasses-json
     httpx-sse
-    langchain
+    langchain-classic
     langchain-core
     langsmith
     numpy

--- a/pkgs/development/python-modules/langchain-core/default.nix
+++ b/pkgs/development/python-modules/langchain-core/default.nix
@@ -36,24 +36,19 @@
 
 buildPythonPackage rec {
   pname = "langchain-core";
-  version = "1.0.0";
+  version = "1.0.7";
   pyproject = true;
 
   src = fetchFromGitHub {
     owner = "langchain-ai";
     repo = "langchain";
     tag = "langchain-core==${version}";
-    hash = "sha256-/V2xwpVIR7AWkN85D51LYokMrTrnJlIkGr0q1ztVh8A=";
+    hash = "sha256-esMMA9z8z9Q909buPGZP1jujEWzdXT+Mxlwvjz0XqKk=";
   };
 
   sourceRoot = "${src.name}/libs/core";
 
   build-system = [ hatchling ];
-
-  pythonRelaxDeps = [
-    "packaging"
-    "tenacity"
-  ];
 
   dependencies = [
     jsonpatch
@@ -74,7 +69,6 @@ buildPythonPackage rec {
     blockbuster
     freezegun
     grandalf
-    httpx
     langchain-tests
     numpy
     pytest-asyncio

--- a/pkgs/development/python-modules/langchain-experimental/default.nix
+++ b/pkgs/development/python-modules/langchain-experimental/default.nix
@@ -4,7 +4,7 @@
   fetchFromGitHub,
 
   # build-system
-  poetry-core,
+  pdm-backend,
 
   # dependencies
   langchain-core,
@@ -20,14 +20,14 @@
 
 buildPythonPackage rec {
   pname = "langchain-experimental";
-  version = "0.3.4";
+  version = "0.4.0";
   pyproject = true;
 
   src = fetchFromGitHub {
     owner = "langchain-ai";
     repo = "langchain-experimental";
     tag = "libs/experimental/v${version}";
-    hash = "sha256-KgGfJfxHOfpwVVo/OcbOjiO5pbxoDE1MiyKqUwsqfIg=";
+    hash = "sha256-A5qCTOCmKt/a1DTKVOC/WwuLCqOYI5pGhAGo/Y4C/FY=";
   };
 
   sourceRoot = "${src.name}/libs/experimental";
@@ -38,7 +38,7 @@ buildPythonPackage rec {
   ];
 
   build-system = [
-    poetry-core
+    pdm-backend
   ];
 
   pythonRelaxDeps = [
@@ -68,7 +68,7 @@ buildPythonPackage rec {
 
   meta = {
     changelog = "https://github.com/langchain-ai/langchain-experimental/releases/tag/${src.tag}";
-    description = "Package add experimental features on LangChain";
+    description = "Experimental features for LangChain";
     homepage = "https://github.com/langchain-ai/langchain-experimental/tree/main/libs/experimental";
     license = lib.licenses.mit;
     maintainers = with lib.maintainers; [ mrdev023 ];

--- a/pkgs/development/python-modules/langchain-experimental/default.nix
+++ b/pkgs/development/python-modules/langchain-experimental/default.nix
@@ -71,6 +71,9 @@ buildPythonPackage rec {
     description = "Experimental features for LangChain";
     homepage = "https://github.com/langchain-ai/langchain-experimental/tree/main/libs/experimental";
     license = lib.licenses.mit;
-    maintainers = with lib.maintainers; [ mrdev023 ];
+    maintainers = with lib.maintainers; [
+      mrdev023
+      sarahec
+    ];
   };
 }

--- a/pkgs/development/python-modules/langchain-groq/default.nix
+++ b/pkgs/development/python-modules/langchain-groq/default.nix
@@ -20,14 +20,14 @@
 
 buildPythonPackage rec {
   pname = "langchain-groq";
-  version = "1.0.0";
+  version = "1.0.1";
   pyproject = true;
 
   src = fetchFromGitHub {
     owner = "langchain-ai";
     repo = "langchain";
     tag = "langchain-groq==${version}";
-    hash = "sha256-OuZEdpPp0mKSejd43RW3bXzCzp3E4Pce7flsSr5JleY=";
+    hash = "sha256-08dEZI+sI33i4FGSAVvE3i5KIRccdslw0PXqXlMLldE=";
   };
 
   sourceRoot = "${src.name}/libs/partners/groq";

--- a/pkgs/development/python-modules/langchain-mistralai/default.nix
+++ b/pkgs/development/python-modules/langchain-mistralai/default.nix
@@ -24,14 +24,14 @@
 
 buildPythonPackage rec {
   pname = "langchain-mistralai";
-  version = "1.0.0";
+  version = "1.0.1";
   pyproject = true;
 
   src = fetchFromGitHub {
     owner = "langchain-ai";
     repo = "langchain";
     tag = "langchain-mistralai==${version}";
-    hash = "sha256-khpZY6kttbgacnY1EKCyIPBR2ZiZHC3OA+0NpIBXg9s=";
+    hash = "sha256-o9xIIcqsuTgWMeluk3EMY3hbB3wGjhYYfzbHizpNTo8=";
   };
 
   sourceRoot = "${src.name}/libs/partners/mistralai";
@@ -46,12 +46,6 @@ buildPythonPackage rec {
     pydantic
   ];
 
-  pythonRelaxDeps = [
-    # Each component release requests the exact latest core.
-    # That prevents us from updating individual components.
-    "langchain-core"
-  ];
-
   nativeCheckInputs = [
     langchain-tests
     pytest-asyncio
@@ -59,6 +53,11 @@ buildPythonPackage rec {
   ];
 
   enabledTestPaths = [ "tests/unit_tests" ];
+
+  disabledTests = [
+    # Comparison error due to message formatting differences
+    "test__convert_dict_to_message_tool_call"
+  ];
 
   pythonImportsCheck = [ "langchain_mistralai" ];
 
@@ -71,7 +70,7 @@ buildPythonPackage rec {
   };
 
   meta = {
-    changelog = "https://github.com/langchain-ai/langchain-mistralai/releases/tag/${src.tag}";
+    changelog = "https://github.com/langchain-ai/langchain/releases/tag/${src.tag}";
     description = "Build LangChain applications with mistralai";
     homepage = "https://github.com/langchain-ai/langchain/tree/master/libs/partners/mistralai";
     license = lib.licenses.mit;

--- a/pkgs/development/python-modules/langchain-mongodb/default.nix
+++ b/pkgs/development/python-modules/langchain-mongodb/default.nix
@@ -30,12 +30,12 @@ buildPythonPackage rec {
 
   src = fetchFromGitHub {
     owner = "langchain-ai";
-    repo = "langchain";
-    tag = "langchain-mongodb==${version}";
+    repo = "langchain-mongodb";
+    tag = "libs/langchain-mongodb/v${version}";
     hash = "sha256-Jd9toXkS9dGtSIrJQ/5W+swV1z2BJOJKBtkyGzj3oSc=";
   };
 
-  sourceRoot = "${src.name}/libs/partners/mongodb";
+  sourceRoot = "${src.name}/libs/langchain-mongodb";
 
   build-system = [ poetry-core ];
 
@@ -70,14 +70,14 @@ buildPythonPackage rec {
     # python updater script sets the wrong tag
     skipBulkUpdate = true;
     updateScript = gitUpdater {
-      rev-prefix = "langchain-mongodb==";
+      rev-prefix = "libs/langchain-mongodb/v";
     };
   };
 
   meta = {
-    changelog = "https://github.com/langchain-ai/langchain/releases/tag/${src.tag}";
+    changelog = "https://github.com/langchain-ai/langchain-mongodb/releases/tag/${src.tag}";
     description = "Integration package connecting MongoDB and LangChain";
-    homepage = "https://github.com/langchain-ai/langchain/tree/master/libs/partners/mongodb";
+    homepage = "https://github.com/langchain-ai/langchain-mongodb";
     license = lib.licenses.mit;
     maintainers = with lib.maintainers; [
       natsukium

--- a/pkgs/development/python-modules/langchain-mongodb/default.nix
+++ b/pkgs/development/python-modules/langchain-mongodb/default.nix
@@ -4,16 +4,24 @@
   fetchFromGitHub,
 
   # build-system
-  poetry-core,
+  hatchling,
 
   # dependencies
+  langchain,
+  langchain-classic,
   langchain-core,
+  langchain-text-splitters,
+  lark,
   numpy,
   pymongo,
 
   freezegun,
   httpx,
-  langchain,
+  langchain-community,
+  langchain-ollama,
+  langchain-openai,
+  langchain-tests,
+  mongomock,
   pytest-asyncio,
   pytestCheckHook,
   pytest-mock,
@@ -25,19 +33,19 @@
 
 buildPythonPackage rec {
   pname = "langchain-mongodb";
-  version = "0.2.0";
+  version = "0.8.0";
   pyproject = true;
 
   src = fetchFromGitHub {
     owner = "langchain-ai";
     repo = "langchain-mongodb";
     tag = "libs/langchain-mongodb/v${version}";
-    hash = "sha256-Jd9toXkS9dGtSIrJQ/5W+swV1z2BJOJKBtkyGzj3oSc=";
+    hash = "sha256-TpEiTQe4nZEhb7yWdm73oKaHGr58Ej3cNeD1sP5pAxA=";
   };
 
   sourceRoot = "${src.name}/libs/langchain-mongodb";
 
-  build-system = [ poetry-core ];
+  build-system = [ hatchling ];
 
   pythonRelaxDeps = [
     # Each component release requests the exact latest core.
@@ -47,7 +55,10 @@ buildPythonPackage rec {
   ];
 
   dependencies = [
+    langchain
+    langchain-classic
     langchain-core
+    langchain-text-splitters
     numpy
     pymongo
   ];
@@ -55,7 +66,12 @@ buildPythonPackage rec {
   nativeCheckInputs = [
     freezegun
     httpx
-    langchain
+    langchain-community
+    langchain-ollama
+    langchain-openai
+    langchain-tests
+    lark
+    mongomock
     pytest-asyncio
     pytestCheckHook
     pytest-mock
@@ -63,6 +79,11 @@ buildPythonPackage rec {
   ];
 
   enabledTestPaths = [ "tests/unit_tests" ];
+
+  disabledTestPaths = [
+    # Expects a MongoDB cluster and are very slow
+    "tests/unit_tests/test_index.py"
+  ];
 
   pythonImportsCheck = [ "langchain_mongodb" ];
 

--- a/pkgs/development/python-modules/langchain-openai/default.nix
+++ b/pkgs/development/python-modules/langchain-openai/default.nix
@@ -33,25 +33,19 @@
 
 buildPythonPackage rec {
   pname = "langchain-openai";
-  version = "1.0.1";
+  version = "1.0.3";
   pyproject = true;
 
   src = fetchFromGitHub {
     owner = "langchain-ai";
     repo = "langchain";
     tag = "langchain-openai==${version}";
-    hash = "sha256-lKZZw9kMV3oM7fNpVvofZJfOcyoUdqByWQmBV5MTFZo=";
+    hash = "sha256-6xJEzwJlKGJg6LBQdKIt1E0dkSJcnocUOXq+NzUVEEc=";
   };
 
   sourceRoot = "${src.name}/libs/partners/openai";
 
   build-system = [ hatchling ];
-
-  pythonRelaxDeps = [
-    # Each component release requests the exact latest core.
-    # That prevents us from updating individual components.
-    "langchain-core"
-  ];
 
   dependencies = [
     langchain-core
@@ -85,11 +79,7 @@ buildPythonPackage rec {
     "test_embed_documents_with_custom_chunk_size"
     "test_get_num_tokens_from_messages"
     "test_get_token_ids"
-    "test_init_o1"
-
-    # TypeError: exceptions must be derived from Warning, not <class 'NoneType'>
-    # https://github.com/langchain-ai/langchain/issues/33705
-    "test_init_minimal_reasoning_effort"
+    "test_embeddings_respects_token_limit"
   ];
 
   pythonImportsCheck = [ "langchain_openai" ];

--- a/pkgs/development/python-modules/langgraph-checkpoint-mongodb/default.nix
+++ b/pkgs/development/python-modules/langgraph-checkpoint-mongodb/default.nix
@@ -1,0 +1,54 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchFromGitHub,
+
+  # build-system
+  hatchling,
+
+  # dependencies
+  langgraph-checkpoint,
+  langchain-mongodb,
+  pymongo,
+}:
+
+buildPythonPackage rec {
+  pname = "langgraph-checkpoint-mongodb";
+  version = "0.3.0";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "langchain-ai";
+    repo = "langchain-mongodb";
+    tag = "libs/langgraph-checkpoint-mongodb/v${version}";
+    hash = "sha256-pfgqJ7QPtAxV86HEEaRBMAOCKC8y/iTlZmPzu1szE/o=";
+  };
+
+  sourceRoot = "${src.name}/libs/langgraph-checkpoint-mongodb";
+
+  build-system = [
+    hatchling
+  ];
+
+  dependencies = [
+    langgraph-checkpoint
+    langchain-mongodb
+    pymongo
+  ];
+
+  enabledTestPaths = [ "tests/unit_tests" ];
+
+  # Connection refused (to localhost:27017) for all tests
+  doCheck = false;
+
+  # no pythonImportsCheck as this package does not provide any direct imports
+  pythonImportsCheck = [ "langgraph.checkpoint.mongodb" ];
+
+  meta = {
+    description = "Integrations between MongoDB, Atlas, LangChain, and LangGraph";
+    homepage = "https://github.com/langchain-ai/langchain-mongodb/tree/main/libs/langgraph-checkpoint-mongodb";
+    changelog = "https://github.com/langchain-ai/langchain-mongodb/releases/tag/${src.tag}";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ sarahec ];
+  };
+}

--- a/pkgs/development/python-modules/langgraph-store-mongodb/default.nix
+++ b/pkgs/development/python-modules/langgraph-store-mongodb/default.nix
@@ -1,0 +1,50 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchFromGitHub,
+
+  # build-system
+  hatchling,
+
+  # dependencies
+  langgraph-checkpoint,
+  langchain-mongodb,
+}:
+
+buildPythonPackage rec {
+  pname = "langgraph-store-mongodb";
+  version = "0.1.1";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "langchain-ai";
+    repo = "langchain-mongodb";
+    tag = "libs/langgraph-store-mongodb/v${version}";
+    hash = "sha256-/dafQK6iy85rMt7FMVb3ssaIop9JjjrwajaHAfBUp7g=";
+  };
+
+  sourceRoot = "${src.name}/libs/langgraph-store-mongodb";
+
+  build-system = [
+    hatchling
+  ];
+
+  dependencies = [
+    langgraph-checkpoint
+    langchain-mongodb
+  ];
+
+  # Connection Refused (to mongodb://localhost:27017)
+  # Note that `langchain-mongodb` runs the same tests with mocks.
+  doCheck = false;
+
+  pythonImportsCheck = [ "langgraph.store.mongodb" ];
+
+  meta = {
+    description = "Integrations between MongoDB, Atlas, LangChain, and LangGraph";
+    homepage = "https://github.com/langchain-ai/langchain-mongodb/tree/main/libs/langgraph-store-mongodb";
+    changelog = "https://github.com/langchain-ai/langchain-mongodb/releases/tag/${src.tag}";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ sarahec ];
+  };
+}

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -8079,6 +8079,8 @@ self: super: with self; {
 
   langchain-chroma = callPackage ../development/python-modules/langchain-chroma { };
 
+  langchain-classic = callPackage ../development/python-modules/langchain-classic { };
+
   langchain-community = callPackage ../development/python-modules/langchain-community { };
 
   langchain-core = callPackage ../development/python-modules/langchain-core { };

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -8139,6 +8139,8 @@ self: super: with self; {
 
   langgraph-sdk = callPackage ../development/python-modules/langgraph-sdk { };
 
+  langgraph-store-mongodb = callPackage ../development/python-modules/langgraph-store-mongodb { };
+
   langid = callPackage ../development/python-modules/langid { };
 
   langsmith = callPackage ../development/python-modules/langsmith { };

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -8123,6 +8123,10 @@ self: super: with self; {
 
   langgraph-checkpoint = callPackage ../development/python-modules/langgraph-checkpoint { };
 
+  langgraph-checkpoint-mongodb =
+    callPackage ../development/python-modules/langgraph-checkpoint-mongodb
+      { };
+
   langgraph-checkpoint-postgres =
     callPackage ../development/python-modules/langgraph-checkpoint-postgres
       { };

